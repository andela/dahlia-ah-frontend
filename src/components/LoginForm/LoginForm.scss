
.login-form {
  .form-error {
    color: red;
    margin: 5px 0;
    display: block;
  }
  
  .loader {
    position: relative !important;
    top: 0 !important;
    right: 0 !important;
  }
  .btn, .btn:hover, .btn:focus {
    color: rgba(0, 0, 0, 0.9);
    background: #ffffff;
    border: 2px solid rgba(0, 0, 0, 0.9);
    border-radius: 0;
    margin-top: 10px;
    padding: 0 30px;
    box-shadow: none;
  }
  
  .login-input {
    margin-bottom: 20px !important;
  }
  
  .login-input:focus, .login-input:focus, .login-input:focus {
    border-bottom: 1px solid rgba(0, 0, 0, 0.9) !important;
    box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.9) !important;
  }
  
  .align-right {
    text-align: right;
    color: #747474;
  }

  .forgot-password {
    background: none;
    border: none;
    padding: 0;
    color: #747474;
    cursor: pointer;
  }

  .btn {
    letter-spacing: 0 !important;
  }
}
