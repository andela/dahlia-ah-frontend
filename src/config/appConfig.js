export default {
  FRONTEND_PATH: 'http://localhost:3000',
  BASE_PATH: 'http://localhost:5000',
  BACKEND_PATH: 'http://localhost:5000/api/v1',
};
