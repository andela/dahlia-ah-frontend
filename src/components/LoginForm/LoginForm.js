import React from 'react';
import { withRouter } from 'react-router-dom';
import validator from 'validator';
import ReactRouterPropTypes from 'react-router-prop-types';
import LoginFormRender from './LoginFormRender';
import loginApiCall from '../../api/login';
import {
  emailState, passwordState, errorState, isLoadingState,
} from '../../hooks/Login';

const LoginForm = ({ history }) => {
  const { email, setEmail } = emailState();
  const { password, setPassword } = passwordState();
  const { error, setError } = errorState();
  const { isLoading, setLoader } = isLoadingState();
  const handleValidation = () => {
    const isValidEmail = validator.isEmail(email);
    if (!isValidEmail) {
      setError('The email you entered is invalid');
      setLoader(false);
    } else {
      setError('');
    }
    return isValidEmail;
  };
  const handleSubmit = (event) => {
    event.preventDefault();
    setLoader(true);
    if (handleValidation()) {
      const formData = { email, password };
      loginApiCall(formData, history, setError, setLoader);
    }
  };
  return (
    <LoginFormRender
      handleSubmit={handleSubmit}
      setEmail={setEmail}
      email={email}
      setPassword={setPassword}
      password={password}
      error={error}
      isLoading={isLoading}
    />
  );
};

LoginForm.propTypes = {
  history: ReactRouterPropTypes.history.isRequired,
};

export default withRouter(LoginForm);
