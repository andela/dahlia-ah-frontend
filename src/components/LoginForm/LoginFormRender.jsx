import React from 'react';
import PropType from 'prop-types';
import './LoginForm.scss';

const LoginFormRender = ({
  handleSubmit, setEmail, email, setPassword, password, error, isLoading, openModal,
}) => (
  <form className="login-form" onSubmit={handleSubmit}>
    <input id="email" className="login-input" type="email" autoComplete="email" name="email" placeholder="Email" onChange={(event) => setEmail(event.target.value)} value={email} required />
    <input className="login-input" type="password" autoComplete="current-password" name="password" placeholder="Password" onChange={(event) => setPassword(event.target.value)} value={password} required />
    <div className="align-right"><button className="forgot-password" onClick={() => openModal('reset-password')} type="button">Forgot Password?</button></div>
    <span className="form-error">{error}</span>
    <button type="submit" className="btn">
      { !isLoading ? 'Sign In' : <i className="fa fa-spinner fa-spin loader" />}
    </button>
  </form>
);

LoginFormRender.propTypes = {
  handleSubmit: PropType.func.isRequired,
  setEmail: PropType.func.isRequired,
  email: PropType.string.isRequired,
  setPassword: PropType.func.isRequired,
  openModal: PropType.func.isRequired,
  password: PropType.string.isRequired,
  error: PropType.string.isRequired,
  isLoading: PropType.bool.isRequired,
};

export default LoginFormRender;
