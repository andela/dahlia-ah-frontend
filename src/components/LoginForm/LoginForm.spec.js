import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { renderHook, act } from '@testing-library/react-hooks';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import appConfig from '../../config/appConfig';
import loginApiCall from '../../api/login';
import LoginForm from './LoginForm';
import LoginFormRender from './LoginFormRender';
import {
  emailState, passwordState, errorState, isLoadingState,
} from '../../hooks/Login';

const { BASE_PATH } = appConfig;

configure({ adapter: new Adapter() });

describe('LoginForm component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <LoginForm isOpen={false} closeModal={() => {}} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
describe('LoginFormRenderer component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <LoginFormRender handleSubmit={() => {}} setEmail={() => {}} email="" setPassword={() => {}} password="" error="" isLoading={false} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('TEST FOR LOGIN HOOKS', () => {
  it('should update email', () => {
    const { result } = renderHook(() => emailState(''));
    act(() => {
      result.current.setEmail('eden@gmail.com');
    });

    expect(result.current.email).toBe('eden@gmail.com');
  });
  it('should update password', () => {
    const { result } = renderHook(() => passwordState(''));
    act(() => {
      result.current.setPassword('password');
    });

    expect(result.current.password).toBe('password');
  });
  it('should update error', () => {
    const { result } = renderHook(() => errorState(''));
    act(() => {
      result.current.setError('error');
    });

    expect(result.current.error).toBe('error');
  });
  it('should update isLoading', () => {
    const { result } = renderHook(() => isLoadingState(''));
    expect(result.current.isLoading).toBe(false);
    act(() => {
      result.current.setLoader(true);
    });
    expect(result.current.isLoading).toBe(true);
  });
});

describe('TEST API CALLS RESPONSE', () => {
  it('should return 200 and update localStorage', () => {
    const mock = new MockAdapter(axios);
    const mockData = { data: { user: { token: 'test' } } };
    const url = `${BASE_PATH}/auth/login`;
    mock.onPost(url).reply(200, mockData);

    localStorage.setItem('AuthorsHavenToken', '');

    loginApiCall({ }, { push: () => {} }, () => {}, () => {});
    expect(localStorage.getItem('AuthorsHavenToken')).toBe('');
  });
  it('should return 200 and update localStorage', () => {
    const mock = new MockAdapter(axios);
    const url = `${BASE_PATH}/auth/login`;
    mock.onPost(url).reply(401);

    localStorage.setItem('AuthorsHavenToken', '');

    loginApiCall({}, { push: () => {} }, () => {}, () => {});
    expect(localStorage.getItem('AuthorsHavenToken')).toBe('');
    localStorage.removeItem('AuthorsHavenToken');
    expect(localStorage.getItem('AuthorsHavenToken')).toBe(null);
  });
});
