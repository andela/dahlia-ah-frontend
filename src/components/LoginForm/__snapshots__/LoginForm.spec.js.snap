// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component renders correctly 1`] = `ShallowWrapper {}`;

exports[`LoginFormRenderer component renders correctly 1`] = `ShallowWrapper {}`;
